File: src/main/java/com/machfour/macros/queries/QueryHelpers.java
package com.machfour.macros.queries;

import com.machfour.macros.core.Column;
import com.machfour.macros.core.Schema;
import com.machfour.macros.core.Table;
import com.machfour.macros.objects.CompositeFood;
import com.machfour.macros.objects.Food;
import com.machfour.macros.objects.FoodCategory;
import com.machfour.macros.objects.FoodPortion;
import com.machfour.macros.objects.FoodType;
import com.machfour.macros.objects.Ingredient;
import com.machfour.macros.objects.Meal;
import com.machfour.macros.objects.NutritionData;
import com.machfour.macros.objects.QtyUnit;
import com.machfour.macros.objects.QtyUnits;
import com.machfour.macros.objects.Serving;
import com.machfour.macros.storage.MacrosDataSource;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import static com.machfour.macros.core.MacrosUtils.getOrDefault;

class QueryHelpers {
    static <M, J> Map<J, Long> getIdsFromKeys(MacrosDataSource ds, Table<M> t, Column<M, J> keyCol, Collection<J> keys) throws SQLException {
        return keys.isEmpty() ? Collections.emptyMap() : ds.getIdsByKeysNoEmpty(t, keyCol, keys);
    }

    // Makes meal objects, filtering by the list of IDs. If mealIds is empty,
    // all meals will be returned.
    @NotNull
    static Map<Long, Meal> getRawMealsById(MacrosDataSource ds, @NotNull List<Long> mealIds) throws SQLException {
        return getRawObjectsByKeys(ds, Meal.table(), Schema.MealTable.ID, mealIds);
    }

    @Nullable
    static <M> M getRawObjectById(MacrosDataSource ds, Table<M> t, Long id) throws SQLException {
        return getRawObjectByKey(ds, t, t.getIdColumn(), id);
    }

    @NotNull
    static <M> Map<Long, M> getRawObjectsByIds(MacrosDataSource ds, Table<M> t, Collection<Long> ids) throws SQLException {
        return getRawObjectsByKeys(ds, t, t.getIdColumn(), ids);
    }

    @Nullable
    static <M, J> M getRawObjectByKey(MacrosDataSource ds, Table<M> t, Column<M, J> keyCol, J key) throws SQLException {
        Map<J, M> returned = getRawObjectsByKeys(ds, t, keyCol, Collections.singletonList(key));
        return getOrDefault(returned, key, null);
    }

    @NotNull
    static <M, J> Map<J, M> getRawObjectsByKeys(MacrosDataSource ds, Table<M> t, Column<M, J> keyCol, Collection<J> keys) throws SQLException {
        return keys.isEmpty() ? Collections.emptyMap() : ds.getRawObjectsByKeysNoEmpty(t, keyCol, keys);
    }

    static void processRawIngredients(MacrosDataSource ds, Map<Long, Ingredient> ingredientMap) throws SQLException {
        List<Long> foodIds = new ArrayList<>(ingredientMap.size());
        List<Long> servingIds = new ArrayList<>(ingredientMap.size());
        for (Ingredient i : ingredientMap.values()) {
            foodIds.add(i.getIngredientFoodId());
            if (i.getServingId() != null) {
                servingIds.add(i.getServingId());
            }
        }
        // XXX make sure this doesn't loop infinitely if two composite foods contain each other as ingredients
        // (or potentially via a longer chain -- A contains B, B contains C, C contains A)
        Map<Long, Food> ingredientFoods = FoodQueries.getFoodsById(ds, foodIds);
        Map<Long, Serving> ingredientServings = FoodQueries.getServingsById(ds, servingIds);

        for (Ingredient i : ingredientMap.values()) {
            // applyFoodsToRawIngredients(ingredients, servings
            Food f = ingredientFoods.get(i.getIngredientFoodId());
            i.setIngredientFood(f);
            // applyServingsToRawIngredients(ingredients, servings)
            if (i.getServingId() != null) {
                Serving s = ingredientServings.get(i.getServingId());
                i.setServing(s);
            }
        }
    }

    static void processRawFoodMap(Map<Long, Food> foods, Map<Long, Serving> servings,
            Map<Long, NutritionData> nutritionData, Map<Long, Ingredient> ingredients,
            Map<String, FoodCategory> categories) {
        applyServingsToRawFoods(foods, servings);
        applyNutritionDataToRawFoods(foods, nutritionData);
        applyIngredientsToRawFoods(foods, ingredients);
        applyFoodCategoriesToRawFoods(foods, categories);
    }

    // foodMap is a map of food IDs to the raw (i.e. unlinked) object created from the database
    static void processRawFoodMap(MacrosDataSource ds, Map<Long, Food> foodMap) throws SQLException {
        if (!foodMap.isEmpty()) {
            //Map<Long, Serving> servings = getRawServingsForFoods(idMap);
            //Map<Long, NutritionData> nData = getRawNutritionDataForFoods(idMap);
            Map<Long, Serving> servings = getRawObjectsForParentFk(ds, foodMap, Serving.table(), Schema.ServingTable.FOOD_ID);
            Map<Long, NutritionData> nutritionData = getRawObjectsForParentFk(ds, foodMap, NutritionData.table(), Schema.NutritionDataTable.FOOD_ID);
            Map<Long, Ingredient> ingredients = getRawObjectsForParentFk(ds, foodMap, Ingredient.table(), Schema.IngredientTable.COMPOSITE_FOOD_ID);
            Map<String, FoodCategory> categories = FoodQueries.getAllFoodCategories(ds);
            processRawIngredients(ds, ingredients);
            processRawFoodMap(foodMap, servings, nutritionData, ingredients, categories);
        }
    }

    @NotNull
    static <M, N> Map<Long, M> getRawObjectsForParentFk(MacrosDataSource ds,
            @NotNull Map<Long, N> parentObjectMap, Table<M> childTable, Column.Fk<M, Long, N> fkCol) throws SQLException {
        Map<Long, M> objectMap = Collections.emptyMap();
        if (!parentObjectMap.isEmpty()) {
            Column<M, Long> childIdCol = childTable.getIdColumn();
            List<Long> ids = Queries.selectColumn(ds, childTable, childIdCol, fkCol, parentObjectMap.keySet());
            if (!ids.isEmpty()) {
                objectMap = getRawObjectsByKeys(ds, childTable, childIdCol, ids);
            } // else no objects in the child table refer to any of the parent objects/rows
        }
        return objectMap;
    }

    static void applyServingsToRawFoods(Map<Long, Food> foodMap, Map<Long, Serving> servingMap) {
        for (Serving s : servingMap.values()) {
            // QtyUnit setup
            QtyUnit unit = QtyUnits.fromAbbreviationNoThrow(s.getQtyUnitAbbr());
            assert (unit != null) : "No quantity unit exists with abbreviation '" + s.getQtyUnitAbbr() + "'";
            s.setQtyUnit(unit);
            // this query should never fail, due to database constraints
            Food f = foodMap.get(s.getFoodId());
            assert (f != null);
            s.setFood(f);
            f.addServing(s);
        }
    }

    static void applyNutritionDataToRawFoods(Map<Long, Food> foodMap, Map<Long, NutritionData> nutritionDataMap) {
        for (NutritionData nd : nutritionDataMap.values()) {
            // this lookup should never fail, due to database constraints
            Food f = foodMap.get(nd.getFoodId());
            assert f != null;
            nd.setFood(f);
            f.setNutritionData(nd);
        }
    }

    // note not all foods in the map will be composite
    static void applyIngredientsToRawFoods(Map<Long, Food> foodMap, Map<Long, Ingredient> ingredientMap) {
        for (Ingredient i : ingredientMap.values()) {
            Food f = foodMap.get(i.getCompositeFoodId());
            assert f instanceof CompositeFood && f.getFoodType() == FoodType.COMPOSITE;
            CompositeFood cf = (CompositeFood) f;
            i.setCompositeFood(cf);
            cf.addIngredient(i);
        }
    }

    static void applyFoodCategoriesToRawFoods(Map<Long, Food> foodMap, Map<String, FoodCategory> categories) {
        for (Food f : foodMap.values()) {
            String categoryName = f.getData(Schema.FoodTable.CATEGORY);
            FoodCategory c = categories.get(categoryName);
            f.setFoodCategory(c);
        }
    }

    /*
     * The map must map the meal ID to the (already created) FoodTable objects needed by FoodPortions
     * in that meal.
     */
    static void applyFoodPortionsToRawMeal(MacrosDataSource ds, Meal meal, Map<Long, Food> foodMap) throws SQLException {
        List<Long> foodPortionIds = Queries.selectColumn(ds, FoodPortion.table(),
                Schema.FoodPortionTable.ID, Schema.FoodPortionTable.MEAL_ID, meal.getId());
        if (!foodPortionIds.isEmpty()) {
            Map<Long, FoodPortion> foodPortions = getRawObjectsByIds(ds, FoodPortion.table(), foodPortionIds);
            for (FoodPortion fp : foodPortions.values()) {
                Food portionFood = foodMap.get(fp.getFoodId());
                assert (portionFood != null);
                fp.setFood(portionFood);
                Long servingId = fp.getServingId();
                if (servingId != null) {
                    Serving serving = portionFood.getServingById(servingId);
                    assert serving != null : "Serving specified by FoodPortion not found in its food!";
                    fp.setServing(serving);
                }
                fp.setMeal(meal);
                meal.addFoodPortion(fp);
            }
        }
    }
}

File: src/main/java/com/machfour/macros/queries/QueryHelpers.kt
package com.machfour.macros.queries

import com.machfour.macros.core.Column
import com.machfour.macros.core.Table
import com.machfour.macros.core.schema.*
import com.machfour.macros.objects.*
import com.machfour.macros.queries.FoodQueries.getAllFoodCategories
import com.machfour.macros.queries.FoodQueries.getFoodsById
import com.machfour.macros.queries.FoodQueries.getServingsById
import com.machfour.macros.storage.MacrosDataSource
import java.sql.SQLException

internal object QueryHelpers {
    @Throws(SQLException::class)
    internal fun <M, J> getIdsFromKeys(ds: MacrosDataSource, t: Table<M>, keyCol: Column<M, J>, keys: Collection<J>): Map<J, Long> {
        return if (keys.isEmpty()) emptyMap() else Queries.getIdsByKeys(ds, t, keyCol, keys)
    }

    // Makes meal objects, filtering by the list of IDs. If mealIds is empty,
    // all meals will be returned.
    @Throws(SQLException::class)
    internal fun getRawMealsById(ds: MacrosDataSource, mealIds: Collection<Long>): Map<Long, Meal> {
        return Queries.getRawObjectsByKeys(ds, Meal.table, MealTable.ID, mealIds)
    }

    @Throws(SQLException::class)
    internal fun <M> getRawObjectById(ds: MacrosDataSource, t: Table<M>, id: Long): M? {
        return getRawObjectByKey(ds, t, t.idColumn, id)
    }

    @Throws(SQLException::class)
    internal fun <M> getRawObjectsByIds(ds: MacrosDataSource, t: Table<M>, ids: Collection<Long>): Map<Long, M> {
        return Queries.getRawObjectsByKeys(ds, t, t.idColumn, ids)
    }

    @Throws(SQLException::class)
    internal fun <M, J> getRawObjectByKey(ds: MacrosDataSource, t: Table<M>, keyCol: Column<M, J>, key: J): M? {
        val returned = Queries.getRawObjectsByKeys(ds, t, keyCol, listOf(key))
        return returned.getOrDefault(key, null)
    }

    @Throws(SQLException::class)
    internal fun processRawIngredients(ds: MacrosDataSource, ingredientMap: Map<Long, Ingredient>) {
        val foodIds = ArrayList<Long>(ingredientMap.size)
        val servingIds = ArrayList<Long>(ingredientMap.size)
        for (i in ingredientMap.values) {
            foodIds.add(i.foodId)
            i.servingId?.let { servingIds += it }
        }
        // XXX make sure this doesn't loop infinitely if two composite foods contain each other as ingredients
        // (or potentially via a longer chain -- A contains B, B contains C, C contains A)
        val ingredientFoods = getFoodsById(ds, foodIds)
        val ingredientServings = getServingsById(ds, servingIds)
        for (i in ingredientMap.values) {
            // applyFoodsToRawIngredients(ingredients, servings
            val f = ingredientFoods.getValue(i.foodId)
            i.initFoodAndNd(f)
            // applyServingsToRawIngredients(ingredients, servings)
            i.servingId?.let { id ->
                val s = ingredientServings.getValue(id)
                i.initServing(s)
            }
        }
    }

    internal fun processRawFoodMap(foods: Map<Long, Food>, servings: Map<Long, Serving>,
                                   nutritionData: Map<Long, FoodNutrientValue>, ingredients: Map<Long, Ingredient>,
                                   categories: Map<String, FoodCategory>) {
        applyServingsToRawFoods(foods, servings)
        applyNutrientValuesToRawFoods(foods, nutritionData)
        applyIngredientsToRawFoods(foods, ingredients)
        applyFoodCategoriesToRawFoods(foods, categories)
    }

    // foodMap is a map of food IDs to the raw (i.e. unlinked) object created from the database
    @Throws(SQLException::class)
    internal fun processRawFoodMap(ds: MacrosDataSource, foodMap: Map<Long, Food>) {
        if (foodMap.isNotEmpty()) {
            //Map<Long, Serving> servings = getRawServingsForFoods(idMap);
            //Map<Long, NutrientData> nData = getRawNutrientDataForFoods(idMap);
            val servings = getRawObjectsForParentFk(ds, foodMap, Serving.table, ServingTable.FOOD_ID)
            val nutrientValues = getRawObjectsForParentFk(ds, foodMap, FoodNutrientValue.table, FoodNutrientValueTable.FOOD_ID)
            val ingredients = getRawObjectsForParentFk(ds, foodMap, Ingredient.table, IngredientTable.PARENT_FOOD_ID)
            val categories: Map<String, FoodCategory> = getAllFoodCategories(ds)
            processRawIngredients(ds, ingredients)
            processRawFoodMap(foodMap, servings, nutrientValues, ingredients, categories)
        }
    }

    @Throws(SQLException::class)
    internal fun <M, N> getRawObjectsForParentFk(
            ds: MacrosDataSource,
            parentObjectMap: Map<Long, N>,
            childTable: Table<M>,
            fkCol: Column.Fk<M, Long, N>
    ): Map<Long, M> {
        var objectMap: Map<Long, M> = emptyMap()
        if (parentObjectMap.isNotEmpty()) {
            val childIdCol = childTable.idColumn
            val ids = Queries.selectSingleColumn(ds, childTable, childIdCol) {
                where(fkCol, parentObjectMap.keys)
            }.map { requireNotNull(it) { "Error: null ID encountered: $it" } }
            if (ids.isNotEmpty()) {
                objectMap = Queries.getRawObjectsByKeys(ds, childTable, childIdCol, ids)
            } // else no objects in the child table refer to any of the parent objects/rows
        }
        return objectMap
    }

    private fun applyServingsToRawFoods(foodMap: Map<Long, Food>, servingMap: Map<Long, Serving>) {
        for (s in servingMap.values) {
            // this query should never fail, due to database constraints
            val f = foodMap.getValue(s.foodId)
            s.initFood(f)
            f.addServing(s)
        }
    }

    private fun applyNutrientValuesToRawFoods(foodMap: Map<Long, Food>, nutrientValueMap: Map<Long, FoodNutrientValue>) {
        for (nv in nutrientValueMap.values) {
            // this lookup should never fail, due to database constraints
            foodMap.getValue(nv.foodId).addNutrientValue(nv)
        }
    }

    // note not all foods in the map will be composite
    private fun applyIngredientsToRawFoods(foodMap: Map<Long, Food>, ingredientMap: Map<Long, Ingredient>) {
        for (i in ingredientMap.values) {
            val f = foodMap[i.parentFoodId]
            require(f is CompositeFood && f.foodType === FoodType.COMPOSITE)
            i.initCompositeFood(f)
            f.addIngredient(i)
        }
    }

    private fun applyFoodCategoriesToRawFoods(foodMap: Map<Long, Food>, categories: Map<String, FoodCategory>) {
        for (f in foodMap.values) {
            val categoryName = f.getData(FoodTable.CATEGORY)
            val c = categories[categoryName]!!
            f.setFoodCategory(c)
        }
    }

    @Throws(SQLException::class)
    internal fun processRawFoodPortions(ds: MacrosDataSource, meal: Meal, fpMap: Map<Long, FoodPortion>, foodMap: Map<Long, Food>) {
        // sort by create time - but if adding to existing meals, they are added to the end
        val fpList = fpMap.values.toList().sortedBy { it.createTime }
        for (fp in fpList) {
            val portionFood = foodMap[fp.foodId]
            require(portionFood != null) { "foodMap did not contain food with ID ${fp.foodId}" }
            fp.initFoodAndNd(portionFood)
            fp.servingId?.let {
                val serving = portionFood.getServingById(it)
                checkNotNull(serving) { "Serving specified by FoodPortion not found in its food!" }
                fp.initServing(serving)
            }
            fp.initMeal(meal)
            meal.addFoodPortion(fp)
        }
    }

    @Throws(SQLException::class)
    private fun getRawFoodPortionsForMeal(ds: MacrosDataSource, meal: Meal): Map<Long, FoodPortion> {
        return Queries.selectSingleColumn(ds, FoodPortion.table, FoodPortionTable.ID) {
            where(FoodPortionTable.MEAL_ID, listOf(meal.id))
            distinct(false)
        }.map { checkNotNull(it) { "Error: null FoodPortion ID encountered: $it" } }
            .takeIf { it.isNotEmpty() }
            ?.let { getRawObjectsByIds(ds, FoodPortion.table, it) }
            ?: emptyMap()
    }

    /*
     * The map must map the meal ID to the (already created) FoodTable objects needed by FoodPortions
     * in that meal.
     */
    @Throws(SQLException::class)
    internal fun applyFoodPortionsToRawMeal(ds: MacrosDataSource, meal: Meal, foodMap: Map<Long, Food>) {
        getRawFoodPortionsForMeal(ds, meal)
            .takeIf { it.isNotEmpty() }
            ?.let { processRawFoodPortions(ds, meal, it, foodMap) }
    }
}